include/ofi_recvwin.h

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2018 Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*
+ * This utility provides a sliding receive window implementation. It uses a
+ * circular queue to order incoming messages based on the message ID in the
+ * transport protocol.
+ */
+
+#if !defined(FI_RECVWIN_H)
+#define FI_RECVWIN_H
+
+#include "config.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <ofi.h>
+#include <ofi_rbuf.h>
+
+#define OFI_DECL_RECVWIN_BUF(entrytype, name)				\
+OFI_DECLARE_CIRQUE(entrytype, recvwin_cirq);				\
+struct name {								\
+	uint64_t exp_msg_id;						\
+	unsigned int win_size;						\
+	struct recvwin_cirq *pending;					\
+};									\
+DECLARE_FREESTACK(struct name, name ##_fs);				\
+									\
+static inline struct name *						\
+ofi_recvwin_buf_alloc(struct name ##_fs *fs, unsigned int size)		\
+{									\
+	struct name *recvq;						\
+	assert(size == roundup_power_of_two(size));			\
+	if (freestack_isempty(fs))					\
+		return NULL;						\
+	recvq		= freestack_pop(fs);				\
+	recvq->exp_msg_id = 0;						\
+	recvq->win_size	= size;						\
+	recvq->pending	= recvwin_cirq_create(recvq->win_size);		\
+	return recvq;							\
+}									\
+									\
+static inline void							\
+ofi_recvwin_free(struct name ##_fs *fs, struct name *recvq)		\
+{									\
+	recvwin_cirq_free(recvq->pending);				\
+	freestack_push(fs, recvq);					\
+}									\
```

Comment:
I would define recvwin init/free (or destroy) calls that take as input a pointer to a recvq structure.  I.e. let the caller allocate the structure where it's convenient, rather than pulling it off a free stack of fixed size.

Reply:
Same comment as above, but yeah it can allow more flexible allocations.